sop-extras/src/Data/SOP/Telescope.hs

```diff
@@ -285,151 +276,78 @@ fromTZ :: Telescope g f '[x] -> f x
 fromTZ (TZ fx)  = fx
 
 {-------------------------------------------------------------------------------
-  Extension and retraction
+  Extension
 -------------------------------------------------------------------------------}
 
-newtype Extend m g f x y = Extend { extendWith :: f x -> m (g x, f y) }
+newtype Extend m g f f' x y = Extend { extendWith :: f x -> m (g x, f' y) }
 
 -- | Extend the telescope
 --
 -- We will not attempt to extend the telescope past its final segment.
-extend :: forall m h g f xs. Monad m
-       => InPairs (Requiring h (Extend m g f)) xs -- ^ How to extend
-       -> NP (f -.-> Maybe :.: h) xs              -- ^ Where to extend /from/
-       -> Telescope g f xs -> m (Telescope g f xs)
-extend = go
+--
+-- We will not attempt to extend the telescope more than one segment.
```

Comment:
This is a significant change.

I think it's justified---at least right now---essentially because the HFC is currently unable to forecast across more than one era boundary. Being able to tick ledger states across multiple era boundaries would technically be possible, but that's not enough for a whole network to keep running. EG Right now even the leadership check relies on forecasting (though it could be changed to use ticking).

It's necessary for this PR, because it allows for the era-transition logic of `extend` (and `align` is similar) to change the type when transitioning. If the function must be able to extend across multiple boundaries, then that argument type must be an endomorphism.

Reply:
FTR: We talked about an idea where `tickLedgerState`/`tickChainDepState` would be split into an "initial" part and an "endomorphism" part, ie (names are horrible)
```haskell
initTicking :: SlotNo ->         LedgerState  -> Ticked (LedgerState)
moreTicking :: SlotNo -> Ticked (LedgerState) -> Ticked (LedgerState)
```
subject to a few laws (group action-like), like `moreTicking b = moreTicking b . moreTicking a` for `b >= a` (which allows us to actually state intuitive notions like "ticking to an intermediate slot and then to the final slot is the same as ticking all the way). (Also, `initTicking` should probably be implemented by ticking by one slot, but that might be a bit annoying with EBBs.)

In any case, if we had sth like this, then the cross-era tick functions could take a *ticked* state as input.